nomos-libp2p/src/behaviour/nat/state_machine/states.rs

```diff
@@ -0,0 +1,210 @@
+use std::fmt::Debug;
+
+use libp2p::Multiaddr;
+
+impl super::State<Uninitialized> {
+    pub const fn new() -> Self {
+        Self {
+            state: Uninitialized(()),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Uninitialized(
+    // Intentionally private so that this state cannot be constructed outside this module
+    (),
+);
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TestIfPublic {
+    addr_to_test: Multiaddr,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TryMapAddress {
+    addr_to_map: Multiaddr,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TestIfMappedPublic {
+    addr_to_test: Multiaddr,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Public {
+    addr: Multiaddr,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MappedPublic {
+    addr: Multiaddr,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Private {
+    addr: Multiaddr,
+}
+
+impl Uninitialized {
```

Comment:
Do you think it would not be better to have impl blocks right after the type they refer to?